prov/util/src/util_attr.c

```diff
@@ -1136,6 +1163,10 @@ void ofi_alter_info(struct fi_info *info, const struct fi_info *hints,
 	for (; info; info = info->next) {
 		info->caps = ofi_get_caps(info->caps, hints->caps, info->caps);
 
+		/* This should stay before call to fi_alter_domain_attr as
+		 * the checks depend on unmodified provider mr_mode attr */
+		info->caps = ofi_mr_mode_caps(info, hints, api_version);
```

Comment:
Setting info->caps twice in a row ends up looking wrong. I'm not sure what the best fix is, but the caps are being manipulated in such a non-obvious way that we need some sort of single modification point.

Reply:
The problem is ofi_mr_mode_caps has to be called only once and before fi_alter_domain_attr while ofi_get_caps gets called multiple times including in fi_alter_domain_attr.

Setting info->caps twice in a row doesn't exactly look good. Agree with you on that. Suggestion: we can create a separate ofi_get_caps function that handles the requirement of each of domain, tx and rx attr.

Reply:
I think we need to do something and see if we can't shake things out to make it cleaner.  You could even try modifying ofi_mr_mode_caps() so that it takes ofi_get_caps() as input, or calls ofi_get_caps() internally.